### PR TITLE
[tt-train] On-device param init

### PR DESCRIPTION
**Pull request by imichalakTT**
### Issue

https://github.com/tenstorrent/tt-metal/issues/39421

### Problem description

Initializing weight and biases for ttml's nanoGPT and Llama takes long time, because of two reasons:

- Models initialize every weight and bias twice, once in the LinearLayer's constructor, the second time in initialize_parameters (or initialize_parameters_gpt2 for nanoGPT).
- Models use numpy (zeros, ones, uniform, normal) to initialize tensors on the CPU, only then move them to the device.

### What's changed

- Add `ttml.ops.randn` that generates tensors with normally-distributed values (C++ and Python).
- Add `ttml.ops.randn` tests.
- Add `ttml.ops.rand` -- tiny wrapper around `ttnn.rand` that sources seeds from ttml's rng.
- Move the Embedding module definitions out of model implementations into single, shared implementation.
- Move the Embedding tests into a separate file.
- Remove `initialize_parameters` and `initialize_parameters_gpt2`.
- Implement following initializers as higher-order functions:
  - `ttml.init.zeros()(shape)`
  - `ttml.init.ones()(shape)`
  - `ttml.init.uniform(low, high)(shape)`
  - `ttml.init.normal(mean, std)(shape)`
- Replace numpy functions with ttnn and ttml equivalents:
  - `np.zeros` -> `ttnn.zeros`
  - `np.ones` -> `ttnn.ones`
  - `np.random.uniform` -> `ttnn.rand`
  - `np.random.normal` -> `ttml.ops.randn`
 - Make LinearLayer and a few other modules accept `weight_init` and `bias_init` as keyword arguments (with defaults matching existing initialization methods).
  
### Results

Reduced `create_model_from_config()` time for tinyllama.yaml (969,369,600 parameters) from 36.518 s to 2.4 s.

### Checklist

No CIs run Python tests, so I ran following locally:

- [x] `tt-train/tests/python/test_embedding.py` (new)
- [x] `tt-train/tests/python/test_randn.py` (new)
- [x] `tt-train/tests/python/test_nanogpt.py`
- [x] `tt-train/tests/python/test_llama.py`

### CODE REVIEW

File: tt-train/sources/ttml/ttml/models/nanogpt/__init__.py [hunk 1]
```diff
@@ -100,29 +67,35 @@ class NanoGPT(AbstractModuleBase):
...
         vocab_size_divisible_by_32 = (config.vocab_size + 31) // 32 * 32
-        self.tok_emb = Embedding(vocab_size_divisible_by_32, config.n_embd)
+        self.tok_emb = Embedding(
+            vocab_size_divisible_by_32,
+            config.n_embd,
+            weight_init=weight_init,
+        )
 
         if config.weight_tying == ttml.models.WeightTyingType.Enabled:
             self.tok_emb.weight = self.fc.weight.tensor
```
[the remark this answers is not in the document]
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/models/nanogpt/__init__.py:82]: known false positive. unrelated to changes in this PR

File: tt-train/sources/ttml/ttml/models/nanogpt/__init__.py [hunk 2]
```diff
@@ -217,25 +176,16 @@ def forward(
...
 
 __all__ = [
-    "Embedding",
     "PositionalEmbedding",
     "TrainablePositionalEmbedding",
     "GPTBlock",
     "NanoGPT",
     "NanoGPTExperimentalConfig",
     "NanoGPTConfig",
     "create_nanogpt",
```
[the remark this answers is not in the document]
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/models/nanogpt/__init__.py:172]: i replaced ttml.models.nanogpt.Embedding imports with ttml.modules.Embedding in existing code (only test_nanogpt.py)

File: tt-train/sources/ttml/ops/rand_op.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace ttml::ops {
+
+namespace {
+// ttnn::rand treats seed==0 as "no seed" (uses random entropy).
+// Shift by +1 to keep all seeds deterministic. UINT32_MAX maps to 1.
+uint32_t avoid_zero_seed(uint32_t seed) {
+    return seed == std::numeric_limits<uint32_t>::max() ? 1 : seed + 1;
+}
+}  // namespace
```
Comment (dyurshevichTT) [inline tt-train/sources/ttml/ops/rand_op.cpp:21]: NIT: shouldn't we just leave 0 seed? Not sure why we want to disregard the random entropy case.

**Review by dyurshevichTT — APPROVED** [review]
Good job, thank you!

File: tt-train/sources/ttml/ops/rand_op.cpp (hunk 3 above)
Comment (imichalakTT) [reply tt-train/sources/ttml/ops/rand_op.cpp:21]: Setting seed to 0 as a way of specifying it should use random entropy is quite awkward in ttnn.rand. So for ttml.ops.rand and ttml.ops.randn seed is std::optional<uint32_t>, if you pass std::nullopt, you get random entropy (seeded from ttml for reproducibility, and not from within ttnn.rand machinery), then 0, 1, and any other uint32_t are stable seeds.

File: tt-train/sources/ttml/ops/randn_op.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+#include "ttnn/operations/eltwise/binary/binary.hpp"
+#include "ttnn/operations/eltwise/unary/unary.hpp"
+#include "ttnn/operations/rand/rand.hpp"
+
+namespace ttml::ops {
+
+namespace {
+// ttnn::rand treats seed==0 as "no seed" (uses random entropy).
+// Shift by +1 to keep all seeds deterministic. UINT32_MAX maps to 1.
+uint32_t avoid_zero_seed(uint32_t seed) {
```
Comment (philei-tt) [inline tt-train/sources/ttml/ops/randn_op.cpp:24]: (nit) Duplication between rand, randn. Can't this just be just one line of code in respective function?

File: tt-train/sources/ttml/ttml/models/llama/gqattn.py [hunk 5]
```diff
@@ -22,6 +22,8 @@ def __init__(
         dropout: float,
         rope_params: ttml.ops.rope.RotaryEmbeddingParams,
         bias_linears: bool = False,
+        weight_init=None,
+        bias_init=None,
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/models/llama/gqattn.py]: I don't like modifying GQA/MHA, Lama block and other unrelated modules, just to init linear layer at the end. I suggest removing those arguments, and just create init in the same place as where you create linear layer

**Review by philei-tt — APPROVED** [review]
Overall great, thank you, just one comment regarding new modules arguments. 

Also, can you update onboarding doc `tt-train/docs/TTML_ONBOARDING.md` with new init functions?

One more thing: do you think you could add pytorch style reinit? Should be quite trivial, and might be useful for people familiar with torch. 
```
layer = nn.Linear(128, 64)
nn.init.uniform_(layer.weight)
nn.init.zeros_(layer.bias)
```

**Review by zbaczewskiTT — APPROVED** [review]

File: tt-train/sources/ttml/ops/rand_op.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    tt::tt_metal::DataType dtype,
+    tt::tt_metal::Layout layout) {
+    auto* device = &autograd::ctx().get_device();
+    ttnn::MemoryConfig mem_config{};
+
+    uint32_t effective_seed = avoid_zero_seed(seed.value_or(autograd::ctx().get_generator()()));
+
+    auto t = ttnn::rand(shape, *device, dtype, layout, mem_config, low, high, effective_seed);
+    return ttml::autograd::create_tensor(t);
+}
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ops/rand_op.cpp:37]: I agree that undocumented 0 as random entropy seed is quite awkward in ttnn. Yet, here you introduce another implicit way of choosing random entropy. 0 is magical in ttnn and std::nullopt is magical in ttml. `avoid_zero_seed` silently changes `ttml::ops::rand` behavior; `ttnn::rand` and `ttml::ops::rand` with `seed=5` will produce different values which I think is surprising. 

What do you think about making it more torch-like? `ttnn::rand` is a general, but `ttml::ops:rand` is specific for training workloads. Maybe we can drop the seed argument altogether, and use context's internal generator for both `ttml::ops::rand()` and `ttml::ops::randn`. This way we get training reproducibility - ops generating random data use context's generator and we drop the awkward seed parameter with implicit randomness on `std::nullopt`.

Context's generator itself can output 0, so we can use:
```
uint32_t make_nonzero_seed(uint32_t seed) {
    return seed == 0 ? 1 : seed;
}
```
To avoid one in a billion scenario.

What do you think?

**Review by zbaczewskiTT — APPROVED** [review]
Great job!

File: tt-train/sources/ttml/ttml/models/llama/gqattn.py (hunk 5 above)
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/models/llama/gqattn.py]: Good idea

File: tt-train/sources/ttml/ops/rand_op.cpp (hunk 6 above)
Comment (imichalakTT) [reply tt-train/sources/ttml/ops/rand_op.cpp:37]: I would say std::nullopt is not magical, or at least less magical than seed=0. std::nullopt is an explicit way of saying I am not passing a seed, let default thing happen, which in this case is a random seed provided from the framework's global rng state, same behavior as torch, exactly what you have described. It's better to think about seed as a "custom_seed" parameter that overrides the default behavior. In torch you can optionally pass your own generator which represents roughly the same concept.

And compared to ttnn.randn it helps that variant is encoded in the type system and function signature and not the value.

> ttnn::rand and ttml::ops::rand with seed=5 will produce different values which I think is surprising.

Maybe a bit surprising, but I am not sure whether matching the exact behavior is worth much.
